Message Structures of the ROS-Industrial Simple Message Protocol

gavanderhoorn opened the pull request:
Progressed sufficiently to request an early review of this.

Note: this is a *retrospective* REP in that it documents the current state of `simple_message` and `industrial_robot_client`, nothing new here, but I got (and answered) enough questions about Simple Message and the entire infrastructure to write it down I believe.

Link to [rendered file](https://github.com/gavanderhoorn/rep-ros-i/blob/a88596b8a12b95a1d3bfb0987fa23389c686199a/rep-ixxxx.rst).

gavanderhoorn committed:
Import latest version of retrospective rep.

gavanderhoorn commented:
@shaun-edwards and @JeremyZoss: as you have been the main authors of `simple_message` and `industrial_robot_client`, could I bother you to take a look and see whether you see anything that is either wrong or missing?

🍔

gavanderhoorn committed:
Fix doc8 issues.

shaun-edwards commented on rep-ixxxx.rst:
```diff
@@ -0,0 +1,826 @@
+#. Refer to `Communication Types`_ for valid values for the ``comm_type`` field.
+#. Refer to `Reply Codes`_ for valid values for the ``reply_code`` field.
+#. For ``TOPIC`` and ``SERVICE_REQUEST`` type messages, the ``reply_code`` field must be set to ``INVALID``.
+#. The ``SUCCESS`` and ``FAILURE`` reply codes shall only be used with ``SERVICE_REPLY`` type messages.
```
This part is a bit ambiguous.  In implementations, I've seen a `FAILURE` behave in multiple ways:
 - indicate the request failed AND no `DATA` payload
 - indicate the request failed AND the `DATA` payload although present, can be ignored.
 - indicate the request failed AND the `DATA` payload contains failure information
 - indicate the request failed AND the `DATA` payload contains a special payload with failure information

To remove all doubt, this should follow the ROS convention, but I'm not entirely sure what the ROS convention is.  A service call returns a boolean, which could be thought of as the `reply_code`.  However, in the event a ROS service returns `false`, I don't believe the reply is populated (even if the service server populated it).

gavanderhoorn replied:
> However, in the event a ROS service returns `false`, I don't believe the reply is populated (even if the service server populated it).

The reply *may* be populated / initialised, but typically only just enough for the error / failure to be communicated (ie: setting an `error_message` or `return_code` field).

As to the four variants you mention: 1 is the most efficient, and I've been using that personally, but only if the semantics of returning a 'header only' message with a `FAILURE` are clear.

`motoman_driver` uses option 3 (or 4, not sure what the difference between those two is).

But is the text really ambiguous? The section describes the `Header` and the acceptable values for each field. Not necessarily how those are used for a particular message.

gavanderhoorn replied:
I'm also slightly confused by your comment.

> A service call returns a boolean, which could be thought of as the `reply_code`. 

to me, `reply_code` is exactly the `bool` you get back when invocation of a ROS service fails.

> However, in the event a ROS service returns `false`, I don't believe the reply is populated (even if the service server populated it).

Ok. So what do you suggest we spec here? The text currently just states that the reply codes `FAILURE` and `SUCCESS` may only be used with messages of type `SERVICE_REPLY`. Nothing more.

Would you like to see a section detailing the semantics of the various ways in which a failure / error can actually be represented / transmitted?

shaun-edwards replied:
Let's just agree that it should mirror a ROS service.  I agree that the `reply_code` should match exactly the ROS Service boolean return.

However, in the event of a failure, the ROS service does not populate the `response`, see [here](http://wiki.ros.org/roscpp/Overview/Services#Callback_Signature), specifically:
>A return value of true means the service succeeded, and the response object has been filled with the necessary data. A return value of false means the call has failed and the response object will not be sent to the caller.

To be consistent, a `reply_code` of `FAILURE` means that the data payload will be empty.  If failure data is required (i.e. an error code), then the `reply_code` will be set to `SUCCESS` and it's up to the caller to parse the response.

This is not the most intuitive, but I feel consistency is better than nothing.

shaun-edwards replied:
I guess I'm advocating for a fixed service reply/response message (like ROS).  Specifically "2. indicate the request failed AND the DATA payload although present, can/should be ignored."

gavanderhoorn replied:
For future readers: addressed this in 372be0663b92ee914d64e0a1ddc0af9e512bbdf1.

shaun-edwards commented on rep-ixxxx.rst:
```diff
@@ -0,0 +1,826 @@
+
+  Prefix
+  Header
+  dummy_data       : shared_real[10]
```
This really exists?  Why does it have dummy data?

gavanderhoorn replied:
Yes. At least the `fanuc_driver` Karel side does this and `abb_driver` Rapid as well.

It's not strictly needed though, so we could spec a header only reply here.

shaun-edwards replied:
Just curious where it originated.  Not sure we can/should change something that has been around so long.

gavanderhoorn replied:
Addressed in eb52fcb95902a5f315a1989c86631a4828077ebc.

shaun-edwards reviewed (changes requested):
Looks pretty complete.  Just one thought and one question.

gavanderhoorn committed:
Explain the context in which this document was written.

gavanderhoorn committed:
Flesh out 'Bytestream layout' section.

gavanderhoorn committed:
Fix formatting.

Single sentence per line, etc.

PetteriAimonen commented:
This is a really useful document, shame it took me quite long to find it! Would be great if it gets merged and linked from the wiki pages.

Just a small note, in the document tri-state values are documented as 0 = ON, 1 = OFF, but in the source code it appears to be the opposite way around:
https://github.com/ros-industrial/industrial_core/blob/kinetic-devel/simple_message/include/simple_message/robot_status.h#L79

gavanderhoorn committed:
Fix tristate enum.

On is actually 1, not 0.

gavanderhoorn commented:
> This is a really useful document, shame it took me quite long to find it! Would be great if it gets merged and linked from the wiki pages.

Yes, that is the plan.

Unfortunately I just haven't had the time to finish this.

I might just link to the PR instead in the mean-time.

> Just a small note, in the document tri-state values are documented as 0 = ON, 1 = OFF, but in the source code it appears to be the opposite way around:
> https://github.com/ros-industrial/industrial_core/blob/kinetic-devel/simple_message/include/simple_message/robot_status.h#L79

Thanks. You're completely correct.

Not sure how that happened.

gavanderhoorn committed:
Show complete bytestream before field dissection.

gavanderhoorn committed:
Finish model-of-operation section

A pretty terse overview of the default model of operation of Simple Message servers and clients.

gavanderhoorn committed:
Clarify what reply_code should be used for

Similar to ROS 1 services, it encodes whether the server was able to handle the service call. NOT whether the service handler was successful (ie: it doesn't reflect whether the action which was performed as a result of invoking the service was successful, but whether the server is capable of handling the requests, which are two different things).

gavanderhoorn committed:
Allow 'short' replies

Lacking the 10 dummy floats.

gavanderhoorn committed:
Clarify what the time fields are for

gavanderhoorn committed:
Clarify units for fields which carry axes related data

Always follow REP 103 (ie: SI units).

gavanderhoorn committed:
Be clear about what's supported by the generic nodes

gavanderhoorn committed:
TODOs have been addressed

Well .. most of them.

gavanderhoorn committed:
Make REP active

gavanderhoorn commented:
Finally finished this.

Addressed all outstanding comments by @shaun-edwards.

CI is green.

Merging, as this is a retrospective REP (essentially protocol documentation).

Further improvements can/will be made in follow-up PRs, if needed.

gavanderhoorn committed:
Update filename

gavanderhoorn committed:
Add REP-I6 to the index

gavanderhoorn commented:
Rebased to use new GHA linter setup.
